src/backend/distributed/commands/foreign_constraint.c

```diff
@@ -34,67 +35,48 @@
 #include "utils/syscache.h"
 
 /* Local functions forward declarations */
-static bool HeapTupleOfForeignConstraintIncludesColumn(HeapTuple heapTuple, Oid
-													   relationId, int pgConstraintKey,
+static bool HeapTupleOfForeignConstraintIncludesColumn(HeapTuple heapTuple,
+													   Oid relationId,
+													   int pgConstraintKey,
 													   char *columnName);
 static void ForeignConstraintFindDistKeys(HeapTuple pgConstraintTuple,
 										  Var *referencingDistColumn,
 										  Var *referencedDistColumn,
 										  int *referencingAttrIndex,
 										  int *referencedAttrIndex);
+static List * GetTableForeignConstraintCommandsInternal(Oid relationOid,
+														AttrNumber attributeNumber,
+														bool excludeSelfReference);
 static Oid get_relation_constraint_oid_compat(HeapTuple heapTuple);
+static List * ForeignKeysToReferenceTables(Oid relationOid, bool findAll);
+static List * GetTableForeignConstraintOidsInternal(Oid relationOid,
+													AttrNumber attributeNumber,
+													bool excludeSelfReference,
+													bool findAll);
 
 /*
  * ConstraintIsAForeignKeyToReferenceTable checks if the given constraint is a
- * foreign key constraint from the given relation to a reference table. It does
- * that by scanning pg_constraint for foreign key constraints.
+ * foreign key constraint from the given relation to a reference table.
  */
 bool
-ConstraintIsAForeignKeyToReferenceTable(char *constraintName, Oid relationId)
+ConstraintIsAForeignKeyToReferenceTable(char *constraintNameInput, Oid relationOid)
 {
-	ScanKeyData scanKey[1];
-	int scanKeyCount = 1;
-	bool foreignKeyToReferenceTable = false;
-
-
-	Relation pgConstraint = heap_open(ConstraintRelationId, AccessShareLock);
-
-	ScanKeyInit(&scanKey[0], Anum_pg_constraint_contype, BTEqualStrategyNumber, F_CHAREQ,
-				CharGetDatum(CONSTRAINT_FOREIGN));
-	SysScanDesc scanDescriptor = systable_beginscan(pgConstraint, InvalidOid, false,
-													NULL, scanKeyCount, scanKey);
+	bool findAll = true;
+	List *foreignKeysToReferenceTables = ForeignKeysToReferenceTables(relationOid,
+																	  findAll);
 
-	HeapTuple heapTuple = systable_getnext(scanDescriptor);
-	while (HeapTupleIsValid(heapTuple))
+	Oid tableForeignConstraintOid = InvalidOid;
```

Comment:
As this variable is very local(used only in the consecutive 3 lines), maybe we can shorten it by renaming to `foreignKeyOid`